File: sudokusolver_scala/src/main/scala/sudokusolver/scala/SudokuSolver.scala
package sudokusolver.scala

import sudokusolver.scala.logic.BruteForceSolution.{MultipleSolutions, NoSolutions, SingleSolution}
import sudokusolver.scala.logic.bruteForce
import sudokusolver.scala.logic.diabolical.*
import sudokusolver.scala.logic.extreme.*
import sudokusolver.scala.logic.simple.*
import sudokusolver.scala.logic.tough.*

import scala.annotation.tailrec

@main def sudokuSolver(board: String): Unit =
  if board.length != UnitSizeSquared || board.exists(!('0' to '9').contains(_)) then
    println(s"board must be $UnitSizeSquared numbers with blanks expressed as 0")
  else
    val message = solve(parseOptionalBoard(board)) match
      case InvalidNoSolutions => "No Solutions"
      case InvalidMultipleSolutions => "Multiple Solutions"
      case Solution(board) => board
      case unableToSolve: UnableToSolve => unableToSolve.message
    println(message)

sealed trait SolveResult

object InvalidNoSolutions extends SolveResult

object InvalidMultipleSolutions extends SolveResult

case class Solution(board: Board[SudokuNumber]) extends SolveResult

class UnableToSolve(board: Board[Cell]) extends SolveResult:
  lazy val message: String =
    s"""
       |Unable to solve:
       |$board
       |
       |Simple String: ${board.toSimpleString}
       |
       |With Candidates:
       |${board.toStringWithCandidates}
       |""".stripMargin

/*
 * The Scala version of solve was written to be purely functional. Unlike the Kotlin or Java implementations, it does
 * not contain any 'var' declarations or mutable collections. Loops have mostly been replaced with tail recursion and
 * pattern matching has been used where applicable.
 */
def solve(input: Board[Option[SudokuNumber]]): SolveResult = bruteForce(input) match
  case NoSolutions => InvalidNoSolutions
  case MultipleSolutions => InvalidMultipleSolutions
  case SingleSolution(bruteForceSolution) =>

    @tailrec
    def solve(board: Board[Cell]): SolveResult =
      if board.cells.collect { case cell: UnsolvedCell => cell }.isEmpty then
        Solution(bruteForceSolution)
      else
        performNextSolution(board) match
          case Seq() => UnableToSolve(board)
          case modifications =>
            val withModifications = modifications.foldLeft(board) { (board, modification) =>
              board(modification.row, modification.column) match
                case SolvedCell(row, column, _) => throw IllegalStateException(s"[$row, $column] is already solved.")
                case UnsolvedCell(row, column, existingCandidates) =>
                  val knownSolution = bruteForceSolution(row, column)
                  val newCell = modification match
                    case RemoveCandidates(_, _, candidatesToRemove) =>
                      for candidate <- candidatesToRemove do
                        if candidate == knownSolution then
                          throw IllegalStateException(s"Cannot remove candidate $candidate from [$row, $column]")
                        if !existingCandidates.contains(candidate) then
                          throw IllegalStateException(s"$candidate is not a candidate of [$row, $column]")
                      UnsolvedCell(row, column, existingCandidates -- candidatesToRemove)
                    case SetValue(_, _, value) =>
                      if value != knownSolution then
                        val message = s"Cannot set value $value to [$row, $column]. Solution is $knownSolution"
                        throw IllegalStateException(message)
                      SolvedCell(row, column, value)
                  board.updated(row, column, newCell)
            }
            solve(withModifications)

    solve(createCellBoard(input))

private def performNextSolution(board: Board[Cell]): Seq[BoardModification] =
  val solutions = LazyList(
    //Start of simple solutions.
    pruneCandidates,
    nakedSingles,
    hiddenSingles,
    nakedPairs,
    nakedTriples,
    hiddenPairs,
    hiddenTriples,
    nakedQuads,
    hiddenQuads,
    pointingPairsPointingTriples,
    boxLineReduction,
    //Start of tough solutions.
    xWing,
    simpleColoringRule2,
    simpleColoringRule4,
    yWing,
    swordfish,
    xyzWing,
    //Start of diabolical solutions.
    xCyclesRule1,
    xCyclesRule2,
    xCyclesRule3,
    bug(_).toSeq,
    xyChains,
    medusaRule1,
    medusaRule2,
    medusaRule3,
    medusaRule4,
    medusaRule5,
    medusaRule6,
    jellyfish,
    uniqueRectanglesType1,
    uniqueRectanglesType2,
    uniqueRectanglesType3,
    uniqueRectanglesType3BWithTriplePseudoCells,
    uniqueRectanglesType4,
    uniqueRectanglesType5,
    extendedUniqueRectangles,
    hiddenUniqueRectangles,
    wxyzWing,
    alignedPairExclusion,
    //Start of extreme solutions.
    groupedXCyclesRule1,
    groupedXCyclesRule2,
    groupedXCyclesRule3,
    emptyRectangles,
    finnedXWing,
    finnedSwordfish,
    alternatingInferenceChainsRule1,
    alternatingInferenceChainsRule2,
    alternatingInferenceChainsRule3,
    sueDeCoq
  )
  solutions.map(_(board)).find(_.nonEmpty).getOrElse(LazyList.empty)

File: SudokuSolver_Swift/SudokuSolver_Swift/SudokuSolver.swift
import ArgumentParser

@main
struct SudokuSolver: ParsableCommand {
    @Argument(help: "Sudoku board as \(unitSizeSquared) numbers with blanks expressed as 0.")
    var board: String
    
    func run() throws {
        if board.count != unitSizeSquared || board.contains(where: { !$0.isASCII || !$0.isWholeNumber }) {
            print("board must be \(unitSizeSquared) numbers with blanks expressed as 0")
        } else {
            do {
                print(try solve(input: Board(optionalBoard: board)))
            } catch BruteForceError.noSolutions {
                print("No Solutions")
            } catch BruteForceError.multipleSolutions {
                print("Multiple Solutions")
            } catch let error as UnableToSolveError {
                print(error.message)
            }
        }
    }
}

struct UnableToSolveError: Error {
    let message: String
    
    init(board: Board<Cell>) {
        message = """
            Unable to solve:
            \(board)
            
            Simple String: \(String(simpleBoard: board))
            
            With Candidates:
            \(String(withCandidates: board))
            """
    }
}

func solve(input: Board<SudokuNumber?>) throws -> Board<SudokuNumber> {
    let bruteForceSolution = try bruteForce(board: input)
    var board = Board(toCellBoard: input)
    var modifications: [BoardModification]
    repeat {
        if board.cells.unsolvedCells.isEmpty {
            return bruteForceSolution
        }
        modifications = performNextSolution(board: board)
        for modification in modifications {
            let row = modification.row
            let column = modification.column
            guard case .unsolvedCell(let cell) = board[row, column] else {
                preconditionFailure("[\(row), \(column) is already solved.")
            }
            let knownSolution = bruteForceSolution[row, column]
            switch modification {
            case .removeCandidates(let modification):
                for candidate in modification.candidates {
                    precondition(
                        candidate != knownSolution,
                        "Cannot remove candidate \(candidate) from [\(row), \(column)]"
                    )
                    precondition(
                        cell.candidates.contains(candidate),
                        "\(candidate) is not a candidate of [\(row), \(column)]"
                    )
                }
                let candidates = cell.candidates.subtracting(modification.candidates)
                board[row, column] = Cell(row: row, column: column, candidates: candidates)
            case .setValue(let modification):
                let value = modification.value
                precondition(
                    value == knownSolution,
                    "Cannot set value \(value) to [\(row), \(column)]. Solution is \(knownSolution)"
                )
                board[row, column] = Cell(row: row, column: column, value: value)
            }
        }
    } while !modifications.isEmpty
    throw UnableToSolveError(board: board)
}

private func performNextSolution(board: Board<Cell>) -> [BoardModification] {
    let solutions = [
        //Start of simple solutions.
        pruneCandidates,
        nakedSingles,
        hiddenSingles,
        nakedPairs,
        nakedTriples,
        hiddenPairs,
        hiddenTriples,
        nakedQuads,
        hiddenQuads,
        pointingPairsPointingTriples,
        boxLineReduction,
        //Start of tough solutions.
        xWing,
        simpleColoringRule2,
        simpleColoringRule4,
        yWing,
        swordfish,
        xyzWing,
        //Start of diabolical solutions.
        xCyclesRule1,
        xCyclesRule2,
        xCyclesRule3,
        { board in bug(board: board).map { [$0] } ?? [] },
        xyChains,
        medusaRule1,
        medusaRule2,
        medusaRule3,
        medusaRule4,
        medusaRule5,
        medusaRule6,
        jellyfish,
        uniqueRectanglesType1,
        uniqueRectanglesType2,
        uniqueRectanglesType3,
        uniqueRectanglesType3BWithTriplePseudoCells,
        uniqueRectanglesType4,
        uniqueRectanglesType5,
        extendedUniqueRectangles,
        hiddenUniqueRectangles,
        wxyzWing,
        alignedPairExclusion,
        //Start of extreme solutions.
        groupedXCyclesRule1,
        groupedXCyclesRule2,
        groupedXCyclesRule3,
        emptyRectangles,
        finnedXWing,
        finnedSwordfish,
        alternatingInferenceChainsRule1,
        alternatingInferenceChainsRule2,
        alternatingInferenceChainsRule3,
        sueDeCoq
    ]
    return solutions.lazy.map { $0(board) }.first { !$0.isEmpty } ?? []
}
